src/dotty/tools/dotc/core/SymbolLoaders.scala

```diff
@@ -86,18 +86,27 @@ class SymbolLoaders {
       (module, modcls) => new PackageLoader(module, pkg)).entered
   }
 
+  /** Whether the class to be loaded is `java.lang.Object`
+   *
+   *  `java.lang.Object` should never be loaded from classfiles
+   */
+  def isObjectClass(owner: Symbol, name: PreName)(implicit ctx: Context): Boolean =
+    owner.fullName.toString == "java.lang" && name.show == "Object"
+
   /** Enter class and module with given `name` into scope of `owner`
    *  and give them `completer` as type.
    */
   def enterClassAndModule(
       owner: Symbol, name: PreName, completer: SymbolLoader,
       flags: FlagSet = EmptyFlags, scope: Scope = EmptyScope)(implicit ctx: Context): Unit = {
     val clazz = enterClass(owner, name, completer, flags, scope)
-    val module = enterModule(
-      owner, name, completer,
-      modFlags = flags.toTermFlags & RetainedModuleValFlags,
-      clsFlags = flags.toTypeFlags & RetainedModuleClassFlags,
-      scope = scope)
+
```

Comment:
It's a way to do it, but not very elegant to test this for every class and module that's entered. 

I played with an alternative strategy, which also seems to work and is less intrusive: IN line 196 of Definitions, where we define ObjectClass, insert:

    val companion = JavaLangPackageVal.info.decl(nme.Object).symbol
    companion.info = NoType // to indicate that it does not really exist